Flaky tests in TestQueueApi::test_api_queue_sort due to size sorting

### SABnzbd version

develop

### Operating system

macOS

### Using Docker image

None

### Description

A note for myself to fix this flaky test sometime.

I've looked at it before and it is an issue with `_create_random_queue` and file sizes close to unit barriers.
I suspect it decided `1024 KB` is equal to `1.0 MB` so the order depends on the queue order or something.
But the main issue is probably how it decides they are not both `1.0 MB`.

```
        # Verify the result
>       assert new_order == original_order
E       AssertionError: assert ['798 KB', '8...'1.2 MB', ...] == ['798 KB', '8...'1.2 MB', ...]
E         
E         At index 3 diff: '1024 KB' != '1.0 MB'
E         
E         Full diff:
E           [
E               '798 KB',
E               '877 KB',
E               '1000 KB',
E         +     '1024 KB',
E               '1.0 MB',
E         -     '1024 KB',
E               '1.2 MB',
E               '1.3 MB',
E               '1.7 MB',
E           ]
```

```
FAILED tests/test_functional_api.py::TestQueueApi::test_api_queue_sort[size-size-asc] - AssertionError: assert ['798 KB', '8...'1.2 MB', ...] == ['798 KB', '8...'1.2 MB', ...]
  
  At index 3 diff: '1024 KB' != '1.0 MB'
  
  Full diff:
    [
        '798 KB',
        '877 KB',
        '1000 KB',
  +     '1024 KB',
        '1.0 MB',
  -     '1024 KB',
        '1.2 MB',
        '1.3 MB',
        '1.7 MB',
    ]
```

--- a/sabnzbd/misc.py
+++ b/sabnzbd/misc.py
@@ -653,25 +653,26 @@ def to_units(val: int | float, postfix="") -> str:
         # Limit it to 5 as the maximum defined index.
         n = min(5, math.trunc(math.log2(val) / 10))
 
-    # Now we scale our value to the appropriate power of 1024
-    # It is written as 2^10n for symmetry with the
-    # selection above.
-    val = val / 2 ** (10 * n)
-
     # Showing the single decimal per doc string
     if n > 1:
         decimals = 1
     else:
         decimals = 0
 
-    # Rounding to the displayed precision can carry the value up to the next unit,
-    # for example 1048575 would be shown as "1024 K" instead of "1.0 M". Move it to
-    # the next unit instead, unless we are already at the maximum defined index.
-    if n < 5 and float(f"{val:.{decimals}f}") >= 1024:
+    # Now we scale our value to the appropriate power of 1024
+    # It is written as 2^10n for symmetry with the
+    # selection above. Round it to the precision we are going to display, so
+    # what we check below is what ends up in the output.
+    val = round(val / 2 ** (10 * n), decimals)
+
+    # That rounding can carry the value up into the next unit, for example 1048575
+    # would be shown as "1024 K" instead of "1.0 M". Move it up a unit instead,
+    # unless we are already at the maximum defined index.
+    if n < 5 and val >= 1024:
         n += 1
-        val = val / 1024
         if n > 1:
             decimals = 1
+        val = round(val / 1024, decimals)
 
     # We might not have anything at all to append
     if n == 0 and postfix == "":

--- a/sabnzbd/misc.py
+++ b/sabnzbd/misc.py
@@ -664,6 +664,15 @@ def to_units(val: int | float, postfix="") -> str:
     else:
         decimals = 0
 
+    # Rounding to the displayed precision can carry the value up to the next unit,
+    # for example 1048575 would be shown as "1024 K" instead of "1.0 M". Move it to
+    # the next unit instead, unless we are already at the maximum defined index.
+    if n < 5 and float(f"{val:.{decimals}f}") >= 1024:
+        n += 1
+        val = val / 1024
+        if n > 1:
+            decimals = 1
+
     # We might not have anything at all to append
     if n == 0 and postfix == "":
         units = ""

--- a/sabnzbd/misc.py
+++ b/sabnzbd/misc.py
@@ -660,8 +660,8 @@ def to_units(val: int | float, postfix="") -> str:
         decimals = 0
 
     # Now we scale our value to the appropriate power of 1024
-    # It is written as 2^10n for symmetry with the
-    # selection above. Round it to the precision we are going to display, so
+    # It is written as 2^10n for symmetry with the selection above.
+    # Round it to the precision we are going to display, so
     # what we check below is what ends up in the output.
     val = round(val / 2 ** (10 * n), decimals)
 

--- a/tests/test_functional_api.py
+++ b/tests/test_functional_api.py
@@ -706,8 +706,14 @@ class TestQueueApi(ApiTestFunctions):
         new_order = list(filter((geriatric_entry).__ne__, new_order))
         original_order = list(filter((geriatric_entry).__ne__, original_order))
 
-        # Verify the result
-        assert new_order == original_order
+        # Verify the result. The api sorts on the exact values, while the slots only
+        # report them rounded to the nearest display unit ("1024 KB" and "1.0 MB" both
+        # mean 1MiB). Comparing at that lower resolution avoids failing on entries that
+        # are indistinguishable in the api output but not to the actual sort.
+        if key:
+            assert [key(entry) for entry in new_order] == [key(entry) for entry in original_order]
+        else:
+            assert new_order == original_order
 
     @pytest.mark.parametrize(
         "queue_size, index_from, index_to, value2_is_nzo_id",

--- a/tests/test_misc.py
+++ b/tests/test_misc.py
@@ -185,6 +185,16 @@ class TestMisc:
         assert "9.8 G" == misc.to_units(1024 * 1024 * 10000)
         assert "1024.0 P" == misc.to_units(1024**6)
 
+        # Values that round up to the next unit should be shown in that unit
+        assert "1023 K" == misc.to_units(1024**2 - 1024)
+        assert "1.0 M" == misc.to_units(1024**2 - 1)
+        assert "-1.0 M" == misc.to_units(-(1024**2 - 1))
+        assert "1.0 MBla" == misc.to_units(1024**2 - 1, postfix="Bla")
+        assert "1023.9 M" == misc.to_units(1024**3 - 1024**2 // 8)
+        assert "1.0 G" == misc.to_units(1024**3 - 1)
+        assert "1.0 T" == misc.to_units(1024**4 - 1)
+        assert "1.0 P" == misc.to_units(1024**5 - 1)
+
     def test_unit_back_and_forth(self):
         assert 100 == misc.from_units(misc.to_units(100))
         assert 1024 == misc.from_units(misc.to_units(1024))

